src/x8-packw/kr-wasmdot.c.in

```diff
@@ -0,0 +1,245 @@
+  assert(kr == ${KR});
+  assert(sr == 1);
+  assert(weights != NULL);
+  assert(packed_weights != NULL);
+
+  const v128_t vone = wasm_i8x16_splat(1);
+  const v128_t vzero = wasm_i32x4_splat(0);
+  XNN_FORCE_REALIZATION(vone);
+  XNN_FORCE_REALIZATION(vzero);
+  ${TYPE}* out = (${TYPE}*) packed_weights;
+  const ${BTYPE}* b = (const ${BTYPE}*) bias;
+  const uint32_t izp = (uint32_t) (params ? (((const struct xnn_qs8_packw_params*) params)->input_zero_point + ${IZP}): ${IZP});
+  v128_t vzeropoint = wasm_i32x4_splat((int32_t) izp);
+
+  do {
+    // NC main loop multiple of ${NR}
+    const ${TYPE}* w0 = (const ${TYPE}*) weights;
+    size_t n = nc;
+    for (;n >= ${NR}; n -= ${NR}) {
+      int32_t* packed_b = (int32_t*) out;
+      if XNN_LIKELY(b != NULL) {
+        $for N in range(0, NR, 4):
+          const v128_t vb${N>>2} = wasm_v128_load(b + ${N});
+          wasm_v128_store(out + ${N * 4}, vb${N>>2});
+        b += ${NR};
+      } else {
+        $for N in range(0, NR, 4):
+          wasm_v128_store(out + ${N * 4}, vzero);
+      }
+      out += ${NR} * sizeof(${BTYPE});
+
+      $for N in range(1, NR):
+        const ${TYPE}* w${N} = w${N-1} + kc;
+
+      $for N in range(0, NR, 2):
+        v128_t vacc${ABC[N:N+2]} = wasm_i32x4_splat(0);
+
+      // KC main loop multiple of ${NR}x${KR}
+      size_t k = kc;
+      for (; k >= ${2 * KR}; k -= ${2 * KR}) {
+        $for N in range(NR):
+          v128_t v${N}_01 = wasm_v128_load(w${N});
+
+        $for N in range(0, NR, 2):
+          v128_t v${ABC[N:N+2]}_0 = wasm_i64x2_shuffle(v${N}_01, v${N+1}_01, 0, 2);
+          v128_t v${ABC[N:N+2]}_1 = wasm_i64x2_shuffle(v${N}_01, v${N+1}_01, 1, 3);
+
+        $for N in range(0, NR, 2):
+          vacc${ABC[N:N+2]} = wasm_i32x4_relaxed_dot_i8x16_i7x16_add(v${ABC[N:N+2]}_0, vone, vacc${ABC[N:N+2]});
+          vacc${ABC[N:N+2]} = wasm_i32x4_relaxed_dot_i8x16_i7x16_add(v${ABC[N:N+2]}_1, vone, vacc${ABC[N:N+2]});
+
+        $for N in range(0, NR, 2):
+          wasm_v128_store(out + ${N * KR}, v${ABC[N:N+2]}_0);
+
+        $for N in range(0, NR, 2):
+          wasm_v128_store(out + ${(N + 8) * KR}, v${ABC[N:N+2]}_1);
+
+        $for N in range(NR):
+          w${N} += ${2 * KR};
+        out += ${2*NR*KR};
+      }
+
+      for (; k >= ${KR}; k -= ${KR}) {
+        $for N in range(0, NR, 2):
+          v128_t v${ABC[N:N+2]} = wasm_v128_load64_splat(w${N});
+          v${ABC[N:N+2]} = wasm_i64x2_shuffle(v${ABC[N:N+2]}, wasm_v128_load64_splat(w${N+1}), 0, 3);
+
+        $for N in range(0, NR, 2):
+          vacc${ABC[N:N+2]} = wasm_i32x4_relaxed_dot_i8x16_i7x16_add(v${ABC[N:N+2]}, vone, vacc${ABC[N:N+2]});
+
+        $for N in range(0, NR, 2):
+          wasm_v128_store(out + ${N * KR}, v${ABC[N:N+2]});
+
+        $for N in range(NR):
+          w${N} += ${KR};
+        out += ${NR*KR};
+      }
+
+      // KC remainder 1..KR-1
+      if (k != 0) {
+        assert(k >= 1 && k <= ${KR-1});
+
+        const v128_t all_one = wasm_i32x4_splat(-1);
```

Comment:
vone preferred for value of 1 or vmask for value of -1

Reply:
Done.